Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp

```diff
@@ -0,0 +1,370 @@
+{
+    Ref vm = context.vm();
+    JSC::JSLockHolder lock(vm);
+
+    auto maybeProtocolComponent = URLPatternUtilities::URLPatternComponent::compile(vm, makeComponentString(), EncodingCallbackType::Protocol, URLPatternUtilities::URLPatternStringOptions { });
+    if (maybeProtocolComponent.hasException())
+        return maybeProtocolComponent.releaseException();
+
+    auto protocolComponent = maybeProtocolComponent.releaseReturnValue();
+    m_protocolMatchesSpecialSchemeFlag = protocolComponent.matchSpecialSchemeProtocol(context);
+
+    return { };
+}
+
+// https://urlpattern.spec.whatwg.org/#change-state
+void URLPatternConstructorStringParser::changeState(URLPatternConstructorStringParserState newState, size_t skip)
+{
+    if (m_state != URLPatternConstructorStringParserState::Init
+        && m_state != URLPatternConstructorStringParserState::Authority
+        && m_state != URLPatternConstructorStringParserState::Done)
+        setInitComponentFromState(m_result, m_state, makeComponentString());
+
+    if (m_state != URLPatternConstructorStringParserState::Init && newState != URLPatternConstructorStringParserState::Done) {
+        // Set init's hostname to empty if conditions are met.
+        static constexpr std::array validStateConditionsForEmptyHostname { URLPatternConstructorStringParserState::Protocol, URLPatternConstructorStringParserState::Authority, URLPatternConstructorStringParserState::Username, URLPatternConstructorStringParserState::Password };
+        static constexpr std::array validNewStateConditionsForEmptyHostname { URLPatternConstructorStringParserState::Port, URLPatternConstructorStringParserState::Pathname, URLPatternConstructorStringParserState::Search, URLPatternConstructorStringParserState::Hash };
+        if (std::find(validStateConditionsForEmptyHostname.begin(), validStateConditionsForEmptyHostname.end(), m_state) != validStateConditionsForEmptyHostname.end()
+            && std::find(validNewStateConditionsForEmptyHostname.begin(), validNewStateConditionsForEmptyHostname.end(), newState) != validNewStateConditionsForEmptyHostname.end()
+            && m_result.hostname.isNull())
+            m_result.hostname = emptyString();
+        // Set init's pathname to empty if conditions are met.
+        static constexpr std::array validStateConditionsForEmptyPathname { URLPatternConstructorStringParserState::Protocol, URLPatternConstructorStringParserState::Authority, URLPatternConstructorStringParserState::Username, URLPatternConstructorStringParserState::Password, URLPatternConstructorStringParserState::Hostname, URLPatternConstructorStringParserState::Port };
+        static constexpr std::array validNewStateConditionsForEmptyPathname { URLPatternConstructorStringParserState::Search, URLPatternConstructorStringParserState::Hash };
+        if (std::find(validStateConditionsForEmptyPathname.begin(), validStateConditionsForEmptyPathname.end(), m_state) != validStateConditionsForEmptyPathname.end()
```

Comment:
You should consider using `std::ranges::find(validStateConditionsForEmptyPathname, m_state)`. Same below